### DISCUSSION TITLE
How to implement a pause state and exit from the paused state in a state machine?

### DISCUSSION THREAD

**Discussion by vkensou** [dotnet-state-machine/stateless#537] [Q&A] [answered]
I want to implement game flow control based on state machine, including the ability to pause and resume the game. The game should also be able to exit while in the paused state. I have defined the game control flow within the 'In Progress' state, which includes events like initialization, pause, resume, and end. However, when transitioning from the 'Pause' state to the 'Begin' state, the 'In Progress' state is not the current state, so the 'End' event within 'In Progress' is not triggered, resulting in resource leakage. What would be the best way to handle this situation?
![未命名绘图 drawio (3)](https://github.com/dotnet-state-machine/stateless/assets/9306197/93a18b3d-c0ee-4ea0-b98a-d9f946b9d2b1)

**Comment by crozone** [comment, marked as answer]
Consider making "In Progress" a superstate that encapsulates "Paused" and "Running". Then you can have another "Stopped" state as the initial state.

The "Start" trigger would transition "Stopped" -> "In Progress". The initial state for "In Progress" would be "Running".

The "Pause" trigger would transition "Running" -> "Paused".

The "Resume" trigger would transition "Paused" -> "Running".

The "End" trigger would transition "In Progress" -> "Stopped". Resource cleanup occurs on entry to "Stopped".

**Reply by vkensou** [reply]
Thank your reply! I'll try this solution.
